src/main/java/ch/njol/skript/expressions/ExprFilter.java

```diff
@@ -115,146 +104,66 @@ protected Object[] get(Event event) {
+			if (specifiedTypePatterns == null)
+				return false;
 
-			for (Pattern pattern : child.getClassInfo().getUserInputPatterns()) {
-				if (pattern.matcher(s).matches())
+			for (Pattern typePattern : specifiedTypePatterns) {
+				if (typePattern.matcher(candidateString).matches()) {
 					return true;
+				}
 			}
 		}
-		return objects.isLoopOf(s); // nothing matched, so we'll rely on the object expression's logic
+		return false;
 	}
 
-	@Name("Filter Input")
-	@Description("Represents the input in a filter expression. " +
-			"For example, if you ran 'broadcast \"something\" and \"something else\" where [input is \"something\"]" +
-			"the condition would be checked twice, using \"something\" and \"something else\" as the inputs.")
-	@Examples("send \"congrats on being staff!\" to all players where [input has permission \"staff\"]")
-	@Since("2.2-dev36")
-	public static class ExprInput<T> extends SimpleExpression<T> {
-
-		static {
-			Skript.registerExpression(ExprInput.class, Object.class, ExpressionType.COMBINED,
-					"input",
-					"%*classinfo% input"
-			);
-		}
-
-		@Nullable
-		private final ExprInput<?> source;
-		private final Class<? extends T>[] types;
-		private final Class<T> superType;
-		@SuppressWarnings("NotNullFieldNotInitialized")
-		private ExprFilter parent;
-		@Nullable
-		private ClassInfo<?> inputType;
-
-		public ExprInput() {
-			this(null, (Class<? extends T>) Object.class);
-		}
 
-		public ExprInput(@Nullable ExprInput<?> source, Class<? extends T>... types) {
-			this.source = source;
-			if (source != null) {
-				this.parent = source.parent;
-				this.inputType = source.inputType;
-				parent.removeChild(source);
-				parent.addChild(this);
-			}
-
-			this.types = types;
-			this.superType = (Class<T>) Utils.getSuperType(types);
-		}
-
-		@Override
-		public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
-			parent = ExprFilter.getParsing();
-
-			if (parent == null)
-				return false;
-
-			parent.addChild(this);
-			inputType = matchedPattern == 0 ? null : ((Literal<ClassInfo<?>>) exprs[0]).getSingle();
-			return true;
-		}
-
-		@Override
-		protected T[] get(Event event) {
-			Object current = parent.getCurrent();
-			if (inputType != null && !inputType.getC().isInstance(current)) {
-				return null;
-			}
-
-			try {
-				return Converters.convert(new Object[]{current}, types, superType);
-			} catch (ClassCastException e1) {
-				return (T[]) Array.newInstance(superType, 0);
-			}
-		}
-
-		public void setParent(ExprFilter parent) {
-			this.parent = parent;
-		}
+	@Override
+	public boolean isLoopOf(String candidateString) {
+		return unfilteredObjects.isLoopOf(candidateString) || matchesAnySpecifiedTypes(candidateString);
+	}
 
-		@Override
-		public <R> Expression<? extends R> getConvertedExpression(Class<R>... to) {
-			return new ExprInput<>(this, to);
-		}
+	public Set<ExprFilterInput<?>> getDependentInputs() {
+		return dependentInputs;
```

Comment:
An add and remove method would be a good addition to this method.

Should this be immutable from this method?